pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java

```diff
@@ -297,65 +355,118 @@ public BlockId getId() {
-    startEntry.remainingPredicates = new HashSet<>(eligibleEqualityPredicatesMap.keySet());
+    startEntry.remainingPredicateColumns = new HashSet<>(predicatesMap.keySet());
+    startEntry.remainingGroupByColumns = new HashSet<>(groupByColumns);
     searchQueue.add(startEntry);
+
     while (!searchQueue.isEmpty()) {
       SearchEntry searchEntry = searchQueue.remove();
       StarTreeIndexNode current = searchEntry.starTreeIndexnode;
-      HashSet<String> remainingColumnsToFilter = searchEntry.remainingPredicates;
-      //check if its leaf
-      if (current.isLeaf()) {
-        //reached leaf
+      HashSet<String> remainingPredicateColumns = searchEntry.remainingPredicateColumns;
+      HashSet<String> remainingGroupByColumns = searchEntry.remainingGroupByColumns;
+      // Check if its leaf
+      if (current.isLeaf() || (remainingPredicateColumns.isEmpty() && remainingGroupByColumns.isEmpty())) {
+        // reached leaf
         matchedEntries.add(searchEntry);
         continue;
       }
-      //find next set of nodes to search
-      String nextDimension = dimensionIndexToNameMapping.inverse().get(current.getChildDimensionName());
-      HashSet<String> newRemainingPredicates = new HashSet<>();
-      newRemainingPredicates.addAll(remainingColumnsToFilter);
-      newRemainingPredicates.remove(nextDimension);
-      if (!(inEligiblePredicatesMap.containsKey(nextDimension)
-          || nonEqualityPredicatesMap.containsKey(nextDimension))) {
-        //check if there is exact match filter on this column
-        int nextValueId;
-        if (eligibleEqualityPredicatesMap.containsKey(nextDimension)) {
-          nextValueId = eligibleEqualityPredicatesMap.get(nextDimension).dictionaryId;
-        } else {
-          nextValueId = StarTreeIndexNode.all();
-        }
-        SearchEntry newEntry = new SearchEntry();
-        newEntry.starTreeIndexnode = current.getChildren().get(nextValueId);
-        if (newEntry.starTreeIndexnode != null) {
-          newEntry.remainingPredicates = newRemainingPredicates;
-          searchQueue.add(newEntry);
-        }
-      } else {
-        //if there is a group by
-        for (Map.Entry<Integer, StarTreeIndexNode> entry : current.getChildren().entrySet()) {
-          if (entry.getKey() != StarTreeIndexNode.all()) {
-            SearchEntry newEntry = new SearchEntry();
-            newEntry.starTreeIndexnode = entry.getValue();
-            newEntry.remainingPredicates = newRemainingPredicates;
-            searchQueue.add(newEntry);
+      // Find next set of nodes to search
+      String nextDimension =
+          dimensionIndexToNameMapping.inverse().get(current.getChildDimensionName());
+
+      HashSet<String> newRemainingPredicateColumns = new HashSet<>();
+      newRemainingPredicateColumns.addAll(remainingPredicateColumns);
+      HashSet<String> newRemainingGroupByColumns = new HashSet<>();
+      newRemainingGroupByColumns.addAll(remainingGroupByColumns);
```

Comment:
Need to create a copy for each node. Not doing so means nodes may have the reference to the same 'Set' which is being modified, and we don't want that.